File: gomoku_instinct/rules/renju.py
"""连珠规则判定（Python 参考实现）。

这是整个项目的规则规范。所有训练信号都源自这里，因此实现优先追求「明显正确」
而非快速；高性能版本在 csrc/ 下，两者由差分测试锁定一致。

规则要点（RIF 标准，自由开局）：

  * 白方无禁手，五连及以上（含长连）即胜。
  * 黑方**恰好**五连才算胜；长连（≥6）、四四、三三均为禁手。
  * 黑方同时成五与触发禁手时，五连优先判黑胜。
  * 禁手点的语义由配置决定：严格 RIF 是「仍可落子但立即判负」。

两处容易写错的地方，单独说明：

**「四」的计数**  ——  RIF 把「四」定义为「一条五格窗口内己方占四格、余一格为空，
补上即成五」。照字面数窗口会出问题：活四 `.XXXX.` 含有两个这样的窗口
（左端补五、右端补五），会被误判成四四禁手，但活四明明是黑方的正常胜着。

真正的判据是：**两个四是否共用同一组四颗子**。

    .XXXX.     两个窗口的黑子集合都是 {1,2,3,4}   → 同一个四 → 活四，合法
    X.XXX.X    窗口黑子集合分别是 {0,2,3,4} 和 {2,3,4,6} → 两个四 → 四四禁手

所以下面按「黑子位置集合」去重来统计四的个数，而不是按窗口或按成五点。
同一组黑子若有两个成五点，那就是活四，计作一个四。

**「活三」的递归定义**  ——  一个三算活三，当且仅当存在一个空点，落子后能成活四，
**且那一手本身不是禁手**。判断那一手是否禁手又要递归地数活三，所以这是个递归定义。
实践中递归深度极少超过 3，但仍设了深度上限做兜底。
"""

from __future__ import annotations

from dataclasses import dataclass

from .board import Board
from .constants import (
    BLACK,
    DIRECTIONS,
    EMPTY,
    LINE_CENTER,
    LINE_RADIUS,
    Forbidden,
    Outcome,
    WHITE,
)


@dataclass(frozen=True)
class MoveJudgment:
    """一手棋落下之后的判定结果。"""

    outcome: Outcome
    forbidden: Forbidden = Forbidden.NONE
    fours: int = 0
    open_threes: int = 0
    longest_run: int = 0

    @property
    def is_forbidden(self) -> bool:
        return self.forbidden != Forbidden.NONE


# ── 一维直线上的基础判定 ────────────────────────────────────────────────────
# 下面这些函数只认「一条直线 + 一个下标」，不关心二维坐标，
# 因此四个方向可以共用同一套逻辑。


def _run_length(line: list[int], idx: int, color: int) -> int:
    """经过 idx 的同色连子长度（含 idx 本身）。"""
    n = len(line)
    left = idx - 1
    while left >= 0 and line[left] == color:
        left -= 1
    right = idx + 1
    while right < n and line[right] == color:
        right += 1
    return right - left - 1


def _completes_five(line: list[int], p: int, color: int, exact: bool) -> bool:
    """在空点 p 落子后是否成五。

    exact=True（黑方）要求恰好五连：长连是禁手，不算成五，因此也不构成「四」。
    exact=False（白方）五连及以上都算。
    """
    line[p] = color
    run = _run_length(line, p, color)
    line[p] = EMPTY
    return run == 5 if exact else run >= 5


def _four_groups(
    line: list[int], idx: int, color: int, exact_five: bool
) -> dict[frozenset[int], set[int]]:
    """统计 idx 这一手在本方向上造成的「四」。

    返回 {四颗子的位置集合: {该组的成五点集合}}。按位置集合去重，是为了让活四
    的两个成五点归到同一个四上（见模块开头的说明）。
    """
    n = len(line)
    groups: dict[frozenset[int], set[int]] = {}

    # 只看包含 idx 的五格窗口 —— 我们数的是「这一手造成的四」。
    for start in range(idx - 4, idx + 1):
        if start < 0 or start + 4 >= n:
            continue
        own: list[int] = []
        gap = -1
        broken = False
        for i in range(start, start + 5):
            v = line[i]
            if v == color:
                own.append(i)
            elif v == EMPTY:
                if gap >= 0:  # 两个及以上空点，不是四
                    broken = True
                    break
                gap = i
            else:  # 对方子或墙
                broken = True
                break
        if broken or len(own) != 4:
            continue
        if not _completes_five(line, gap, color, exact_five):
            continue
        groups.setdefault(frozenset(own), set()).add(gap)

    return groups


def _count_fours(
    line: list[int], idx: int, color: int, exact_five: bool
) -> tuple[int, bool]:
    """返回 (四的个数, 是否含活四)。"""
    groups = _four_groups(line, idx, color, exact_five)
    return len(groups), any(len(points) >= 2 for points in groups.values())


# ── 规则判定器 ──────────────────────────────────────────────────────────────


class RenjuRules:
    """连珠规则判定器。

    参数对应 configs/rules.yaml；关闭全部禁手即退化为自由五子棋（≥5 连即胜），
    用于小棋盘快速验证与消融实验。
    """

    def __init__(
        self,
        *,
        forbidden_enabled: bool = True,
        overline: bool = True,
        double_four: bool = True,
        double_three: bool = True,
        five_overrides_forbidden: bool = True,
        white_overline_wins: bool = True,
        recursion_depth: int = 64,
    ) -> None:
        self.forbidden_enabled = forbidden_enabled
        self.overline = overline
        self.double_four = double_four
        self.double_three = double_three
        self.five_overrides_forbidden = five_overrides_forbidden
        self.white_overline_wins = white_overline_wins
        self.recursion_depth = recursion_depth
        # 递归封顶发生的次数。应当恒为 0；非零说明深度上限设小了。
        self.depth_exceeded = 0
        # 实际达到过的最大递归深度，用来验证上限是否合理。
        self.max_depth_seen = 0

    @classmethod
    def from_config(cls, cfg: dict) -> "RenjuRules":
        rules = cfg.get("rules", cfg)
        forb = rules.get("forbidden", {})
        return cls(
            forbidden_enabled=forb.get("enabled", True),
            overline=forb.get("overline", True),
            double_four=forb.get("double_four", True),
            double_three=forb.get("double_three", True),
            five_overrides_forbidden=forb.get("five_overrides_forbidden", True),
            white_overline_wins=rules.get("white_overline_wins", True),
            recursion_depth=forb.get("recursion_depth", 64),
        )

    # ── 对外接口 ────────────────────────────────────────────────────────────
    def judge(self, board: Board, r: int, c: int, color: int) -> MoveJudgment:
        """判定在空点 (r, c) 落 color 子的结果。调用后棋盘保持原样。"""
        if board.cell(r, c) != EMPTY:
            raise ValueError(f"({r}, {c}) 不是空点")
        board.set(r, c, color)
        try:
            return self._judge_placed(board, r, c, color, depth=0)
        finally:
            board.set(r, c, EMPTY)

    def is_forbidden(self, board: Board, r: int, c: int) -> bool:
        """(r, c) 对黑方是否为禁手点。"""
        if not self.forbidden_enabled:
            return False
        return self.judge(board, r, c, BLACK).is_forbidden

    def forbidden_map(self, board: Board) -> list[bool]:
        """全盘禁手点标记，长度 size*size。用作辅助监督标签与人类提示。"""
        size = board.size
        out = [False] * (size * size)
        if not self.forbidden_enabled:
            return out
        for idx in range(size * size):
            if board.grid[idx] != EMPTY:
                continue
            r, c = divmod(idx, size)
            out[idx] = self.judge(board, r, c, BLACK).is_forbidden
        return out

    def judge_all(self, board: Board, color: int) -> list[MoveJudgment | None]:
        """对每个空点做一次完整判定；非空点为 None。

        与 C++ 侧 Rules.judge_all 一一对应，是差分测试的比对单位。
        """
        size = board.size
        out: list[MoveJudgment | None] = [None] * (size * size)
        for idx in range(size * size):
            if board.grid[idx] != EMPTY:
                continue
            r, c = divmod(idx, size)
            out[idx] = self.judge(board, r, c, color)
        return out

    # ── 内部实现 ────────────────────────────────────────────────────────────
    def _judge_placed(
        self, board: Board, r: int, c: int, color: int, depth: int
    ) -> MoveJudgment:
        """棋子已经放在盘上时的判定。"""
        if depth > self.max_depth_seen:
            self.max_depth_seen = depth
        lines = [board.line(r, c, dr, dc, LINE_RADIUS) for dr, dc in DIRECTIONS]
        runs = [_run_length(line, LINE_CENTER, color) for line in lines]
        longest = max(runs)

        if color == WHITE:
            wins = longest >= 5 if self.white_overline_wins else longest == 5
            return MoveJudgment(
                Outcome.WHITE_WIN if wins else Outcome.ONGOING, longest_run=longest
            )

        # 关闭禁手即退化为自由五子棋，黑白规则对称。
        if not self.forbidden_enabled:
            wins = longest >= 5 if self.white_overline_wins else longest == 5
            return MoveJudgment(
                Outcome.BLACK_WIN if wins else Outcome.ONGOING, longest_run=longest
            )

        has_five = any(run == 5 for run in runs)
        has_overline = longest >= 6

        # 五连优先：同时成五与触发禁手时判黑胜。
        if has_five and self.five_overrides_forbidden:
            return MoveJudgment(Outcome.BLACK_WIN, longest_run=longest)

        if self.overline and has_overline:
            return MoveJudgment(
                Outcome.WHITE_WIN, Forbidden.OVERLINE, longest_run=longest
            )

        total_fours = 0
        for line in lines:
            count, _ = _count_fours(line, LINE_CENTER, BLACK, exact_five=True)
            total_fours += count
        if self.double_four and total_fours >= 2:
            return MoveJudgment(
                Outcome.WHITE_WIN,
                Forbidden.DOUBLE_FOUR,
                fours=total_fours,
                longest_run=longest,
            )

        open_threes = 0
        if self.double_three:
            for line, (dr, dc) in zip(lines, DIRECTIONS):
                # 已经成四的方向按最高等级归类为四，不再计作三。
                count, _ = _count_fours(line, LINE_CENTER, BLACK, exact_five=True)
                if count:
                    continue
                if self._has_open_three(board, r, c, line, dr, dc, depth):
                    open_threes += 1
            if open_threes >= 2:
                return MoveJudgment(
                    Outcome.WHITE_WIN,
                    Forbidden.DOUBLE_THREE,
                    fours=total_fours,
                    open_threes=open_threes,
                    longest_run=longest,
                )

        if has_five:  # five_overrides_forbidden 关闭时走到这里
            return MoveJudgment(Outcome.BLACK_WIN, longest_run=longest)

        return MoveJudgment(
            Outcome.ONGOING,
            fours=total_fours,
            open_threes=open_threes,
            longest_run=longest,
        )

    def _has_open_three(
        self,
        board: Board,
        r: int,
        c: int,
        line: list[int],
        dr: int,
        dc: int,
        depth: int,
    ) -> bool:
        """本方向上是否形成活三。

        活三 = 存在一个空点，落子后在本方向成活四，且那一手本身不是禁手。
        """
        n = len(line)
        for p in range(max(0, LINE_CENTER - 4), min(n, LINE_CENTER + 5)):
            if line[p] != EMPTY:
                continue

            line[p] = BLACK
            run = _run_length(line, p, BLACK)
            _, makes_open_four = _count_fours(line, p, BLACK, exact_five=True)
            line[p] = EMPTY

            if run >= 5:
                # 该点直接成五或长连，说明原形已经是四，不是三。
                continue
            if not makes_open_four:
                continue

            if depth >= self.recursion_depth:
                # 递归封顶：按「那一手可下」处理，即认定为活三。
                # 正常对局中不会触发，计数供审计。
                self.depth_exceeded += 1
                return True

            pr = r + (p - LINE_CENTER) * dr
            pc = c + (p - LINE_CENTER) * dc
            board.set(pr, pc, BLACK)
            try:
                sub = self._judge_placed(board, pr, pc, BLACK, depth + 1)
            finally:
                board.set(pr, pc, EMPTY)

            if not sub.is_forbidden:
                return True

        return False

File: csrc/renju.cpp
#include "renju.h"

#include <cstring>
#include <stdexcept>

#include "geometry.h"

namespace gi {

namespace {

// 一条直线在某个颜色视角下的位掩码视图。
// 线外的格子不存在于掩码中 —— 窗口必须整体落在 [0, len) 内，越界因此自动排除。
struct LineView {
  uint32_t own = 0;      // 己方子
  uint32_t blocked = 0;  // 对方子（线外不参与）
  int len = 0;
  int pos = 0;   // 当前落点在线内的下标
  int line = 0;  // 线号，用于映射回格子编号
};

inline LineView make_line(const uint8_t* grid, const Geometry& g, int cell,
                          int dir, uint8_t color) {
  LineView lv;
  lv.line = g.line_id[dir][cell];
  lv.pos = g.line_pos[dir][cell];
  const std::vector<int16_t>& cells = g.lines[dir][lv.line];
  lv.len = static_cast<int>(cells.size());
  for (int j = 0; j < lv.len; ++j) {
    const uint8_t v = grid[cells[j]];
    if (v == color) {
      lv.own |= (1u << j);
    } else if (v != EMPTY) {
      lv.blocked |= (1u << j);
    }
  }
  return lv;
}

// 经过 pos 的同色连子长度。
inline int run_length(uint32_t own, int len, int pos) {
  int l = pos;
  while (l > 0 && ((own >> (l - 1)) & 1u)) --l;
  int r = pos;
  while (r + 1 < len && ((own >> (r + 1)) & 1u)) ++r;
  return r - l + 1;
}

// 在空点 gap 补一子是否成五。exact=true（黑方）要求恰好五连：
// 长连是禁手，既不算成五，也就不构成「四」。
inline bool completes_five(uint32_t own, int len, int gap, bool exact) {
  const uint32_t filled = own | (1u << gap);
  const int run = run_length(filled, len, gap);
  return exact ? (run == 5) : (run >= 5);
}

// 统计 pos 这一手在本方向造成的「四」的个数，**按四颗子的位掩码去重**。
//
// 去重是关键：活四 `.XXXX.` 的两个成五点共用同一组四颗子，只算一个四（合法）；
// 而 `X.XXX.X` 的两个四分别由 {0,2,3,4} 与 {2,3,4,6} 构成，是两个独立的四，
// 构成四四禁手。照窗口数或照成五点数都会把这两种情况混为一谈。
inline int count_fours(uint32_t own, uint32_t blocked, int len, int pos,
                       bool exact, bool* open_four) {
  uint32_t keys[5];
  int nkeys = 0;
  bool open = false;

  const int lo = pos - 4 < 0 ? 0 : pos - 4;
  for (int s = lo; s <= pos; ++s) {
    if (s + 4 >= len) break;  // s 递增，之后的窗口只会更靠右
    const uint32_t win = 0x1Fu << s;
    if (blocked & win) continue;
    const uint32_t mine = own & win;
    if (__builtin_popcount(mine) != 4) continue;

    // 窗口内既非己方子也非对方子的那一格必然为空。
    const uint32_t gapmask = win & ~mine;
    const int gap = __builtin_ctz(gapmask);
    if (!completes_five(own, len, gap, exact)) continue;

    int k = -1;
    for (int i = 0; i < nkeys; ++i) {
      if (keys[i] == mine) {
        k = i;
        break;
      }
    }
    if (k < 0) {
      keys[nkeys++] = mine;
    } else {
      // 同一组子落在两个窗口里，说明有两个不同的成五点 —— 这就是活四。
      open = true;
    }
  }
  if (open_four != nullptr) *open_four = open;
  return nkeys;
}

// 某个方向上、落子之后形成的最高棋型等级。
//
// 「活三」在这里用非递归判定（只看本方向能否一手成活四）：它是特征标签而非规则判定，
// 精确的三三禁手仍由 has_open_three 的递归版本负责。
uint8_t level_in_direction(uint32_t own, uint32_t blocked, int len, int pos,
                           bool black) {
  const int run = run_length(own, len, pos);
  if (run >= 6) {
    return static_cast<uint8_t>(black ? Level::OVERLINE : Level::FIVE);
  }
  if (run == 5) return static_cast<uint8_t>(Level::FIVE);

  const bool exact = black;  // 黑方必须恰好五连
  bool open_four = false;
  const int fours = count_fours(own, blocked, len, pos, exact, &open_four);
  if (fours > 0) {
    return static_cast<uint8_t>(open_four ? Level::OPEN_FOUR : Level::FOUR);
  }

  // 试探邻近空点：能一手成活四即为活三；只能成四则为眠三。
  bool can_make_four = false;
  const int lo = pos - 4 < 0 ? 0 : pos - 4;
  const int hi = pos + 4 >= len ? len - 1 : pos + 4;
  for (int p = lo; p <= hi; ++p) {
    const uint32_t bit = 1u << p;
    if ((own & bit) || (blocked & bit)) continue;
    const uint32_t own2 = own | bit;
    if (run_length(own2, len, p) >= 5) continue;  // 那是四，不是三
    bool o4 = false;
    const int f = count_fours(own2, blocked, len, p, exact, &o4);
    if (o4) return static_cast<uint8_t>(Level::OPEN_THREE);
    if (f > 0) can_make_four = true;
  }
  return static_cast<uint8_t>(can_make_four ? Level::CLOSED_THREE : Level::NONE);
}

}  // namespace

void Rules::pattern_map(const uint8_t* grid, int size, uint8_t color,
                        uint8_t* out) const {
  const int n = size * size;
  std::memset(out, 0, static_cast<size_t>(n));

  uint8_t buf[MAX_SIZE * MAX_SIZE];
  std::memcpy(buf, grid, static_cast<size_t>(n));
  const Geometry& g = geometry(size);
  const bool black = (color == BLACK);

  for (int i = 0; i < n; ++i) {
    if (buf[i] != EMPTY) continue;
    buf[i] = color;
    uint8_t best = 0;
    for (int d = 0; d < NUM_DIRS; ++d) {
      const LineView lv = make_line(buf, g, i, d, color);
      const uint8_t level =
          level_in_direction(lv.own, lv.blocked, lv.len, lv.pos, black);
      if (level > best) best = level;
    }
    buf[i] = EMPTY;
    out[i] = best;
  }
}

Judgment Rules::judge_placed(uint8_t* grid, int size, int move, uint8_t color,
                             int depth) const {
  // 记录实际达到的最大递归深度（松散更新即可，只用于审计）。
  int seen = max_depth_.load(std::memory_order_relaxed);
  while (depth > seen &&
         !max_depth_.compare_exchange_weak(seen, depth,
                                           std::memory_order_relaxed)) {
  }

  const Geometry& g = geometry(size);

  LineView lv[NUM_DIRS];
  int runs[NUM_DIRS];
  int longest = 0;
  for (int d = 0; d < NUM_DIRS; ++d) {
    lv[d] = make_line(grid, g, move, d, color);
    runs[d] = run_length(lv[d].own, lv[d].len, lv[d].pos);
    if (runs[d] > longest) longest = runs[d];
  }

  Judgment j;
  j.longest_run = static_cast<uint8_t>(longest);

  if (color == WHITE) {
    const bool wins = cfg_.white_overline_wins ? (longest >= 5) : (longest == 5);
    j.outcome = wins ? Outcome::WHITE_WIN : Outcome::ONGOING;
    return j;
  }

  // 关闭禁手即退化为自由五子棋，黑白规则对称。
  if (!cfg_.forbidden_enabled) {
    const bool wins = cfg_.white_overline_wins ? (longest >= 5) : (longest == 5);
    j.outcome = wins ? Outcome::BLACK_WIN : Outcome::ONGOING;
    return j;
  }

  bool has_five = false;
  for (int d = 0; d < NUM_DIRS; ++d) {
    if (runs[d] == 5) has_five = true;
  }
  const bool has_overline = longest >= 6;

  // 五连优先于禁手。
  if (has_five && cfg_.five_overrides_forbidden) {
    j.outcome = Outcome::BLACK_WIN;
    return j;
  }
  if (cfg_.overline && has_overline) {
    j.outcome = Outcome::WHITE_WIN;
    j.forbidden = Forbidden::OVERLINE;
    return j;
  }

  int total_fours = 0;
  bool dir_has_four[NUM_DIRS];
  for (int d = 0; d < NUM_DIRS; ++d) {
    bool open_four = false;
    const int cnt = count_fours(lv[d].own, lv[d].blocked, lv[d].len, lv[d].pos,
                                /*exact=*/true, &open_four);
    dir_has_four[d] = cnt > 0;
    total_fours += cnt;
  }
  j.fours = static_cast<uint8_t>(total_fours);
  if (cfg_.double_four && total_fours >= 2) {
    j.outcome = Outcome::WHITE_WIN;
    j.forbidden = Forbidden::DOUBLE_FOUR;
    return j;
  }

  int open_threes = 0;
  if (cfg_.double_three) {
    for (int d = 0; d < NUM_DIRS; ++d) {
      // 按最高等级归类：已经成四的方向不再计作三。
      if (dir_has_four[d]) continue;
      if (has_open_three(grid, size, move, d, depth)) ++open_threes;
    }
    j.open_threes = static_cast<uint8_t>(open_threes);
    if (open_threes >= 2) {
      j.outcome = Outcome::WHITE_WIN;
      j.forbidden = Forbidden::DOUBLE_THREE;
      return j;
    }
  }

  if (has_five) {  // five_overrides_forbidden 关闭时走到这里
    j.outcome = Outcome::BLACK_WIN;
    return j;
  }
  j.outcome = Outcome::ONGOING;
  return j;
}

bool Rules::has_open_three(uint8_t* grid, int size, int move, int dir,
                           int depth) const {
  const Geometry& g = geometry(size);
  const LineView lv = make_line(grid, g, move, dir, BLACK);
  const std::vector<int16_t>& cells = g.lines[dir][lv.line];

  // 能把三变成活四的那一手，必定与本手同处一个 4 连之内，距离不超过 3；
  // 取 ±4 是留一格余量。
  const int lo = lv.pos - 4 < 0 ? 0 : lv.pos - 4;
  const int hi = lv.pos + 4 >= lv.len ? lv.len - 1 : lv.pos + 4;

  for (int p = lo; p <= hi; ++p) {
    const uint32_t bit = 1u << p;
    if ((lv.own & bit) || (lv.blocked & bit)) continue;

    const uint32_t own2 = lv.own | bit;
    // 该点直接成五或长连，说明原形已经是四，不是三。
    if (run_length(own2, lv.len, p) >= 5) continue;

    bool open_four = false;
    count_fours(own2, lv.blocked, lv.len, p, /*exact=*/true, &open_four);
    if (!open_four) continue;

    if (depth >= cfg_.recursion_depth) {
      depth_exceeded_.fetch_add(1, std::memory_order_relaxed);
      return true;  // 深度封顶：按「那一手可下」处理
    }

    // 递归：只有当「造活四的那一手本身不是禁手」时，这才是真活三。
    const int cell = cells[p];
    grid[cell] = BLACK;
    const Judgment sub = judge_placed(grid, size, cell, BLACK, depth + 1);
    grid[cell] = EMPTY;
    if (!sub.is_forbidden()) return true;
  }
  return false;
}

Judgment Rules::judge(const uint8_t* grid, int size, int move,
                      uint8_t color) const {
  const int n = size * size;
  if (move < 0 || move >= n) {
    throw std::invalid_argument("move out of range");
  }
  if (grid[move] != EMPTY) {
    throw std::invalid_argument("move is not an empty point");
  }
  uint8_t buf[MAX_SIZE * MAX_SIZE];
  std::memcpy(buf, grid, static_cast<size_t>(n));
  buf[move] = color;
  return judge_placed(buf, size, move, color, 0);
}

void Rules::forbidden_map(const uint8_t* grid, int size, uint8_t* out) const {
  const int n = size * size;
  std::memset(out, 0, static_cast<size_t>(n));
  if (!cfg_.forbidden_enabled) return;

  uint8_t buf[MAX_SIZE * MAX_SIZE];
  std::memcpy(buf, grid, static_cast<size_t>(n));
  for (int i = 0; i < n; ++i) {
    if (buf[i] != EMPTY) continue;
    buf[i] = BLACK;
    const Judgment j = judge_placed(buf, size, i, BLACK, 0);
    buf[i] = EMPTY;
    out[i] = j.is_forbidden() ? 1 : 0;
  }
}

void Rules::judge_all(const uint8_t* grid, int size, uint8_t color,
                      uint8_t* out) const {
  const int n = size * size;
  std::memset(out, 255, static_cast<size_t>(5 * n));

  uint8_t buf[MAX_SIZE * MAX_SIZE];
  std::memcpy(buf, grid, static_cast<size_t>(n));
  for (int i = 0; i < n; ++i) {
    if (buf[i] != EMPTY) continue;
    buf[i] = color;
    const Judgment j = judge_placed(buf, size, i, color, 0);
    buf[i] = EMPTY;
    uint8_t* slot = out + 5 * i;
    slot[0] = static_cast<uint8_t>(j.outcome);
    slot[1] = static_cast<uint8_t>(j.forbidden);
    slot[2] = j.fours;
    slot[3] = j.open_threes;
    slot[4] = j.longest_run;
  }
}

}  // namespace gi
